tests/FakeItEasy.IntegrationTests/DummyTests.cs

```diff
@@ -45,10 +47,68 @@ public void Dummy_factories_are_polled_only_once_per_dummy_type()
                 .Be(1, "the factory should've been cached after creating the first Dummy");
         }
 
+        [Fact]
+        public void Same_dummy_type_can_be_created_concurrently_on_different_threads()
+        {
+            // Simulate concurrent dummy type creation using dummy factories
+            // and manual reset events. Ensures that the DummyA instance
+            // requested inside the DummyBDummyFactory is created while
+            // the first DummyA is still being created.
+            // Guards against the concurrent creations being detected as a
+            // circular dependency and throwing an exception.
+            var makeOne = Task.Factory.StartNew(A.Dummy<DummyA>);
+            var makeTwo = Task.Factory.StartNew(A.Dummy<DummyB>);
+
+            var exception = Record.Exception(() => Task.WaitAll(makeOne, makeTwo));
+
+            exception.Should().BeNull("all creations should have succeeded");
+        }
+
         public class Dummy
         {
         }
 
+        [SuppressMessage("Microsoft.Design", "CA1053:StaticHolderTypesShouldNotHaveConstructors", Justification = "Must not be static to make a Dummy.")]
+        public sealed class DummyA
+        {
+            public DummyA()
+            {
+                if (CreatingLock.IsSet)
+                {
+                    return;
+                }
+
+                CreatingLock.Set();
+                DummyBDummyFactory.CreatedLock.Wait();
+            }
+
+            public static ManualResetEventSlim CreatingLock { get; } = new ManualResetEventSlim(false);
+        }
+
+        public sealed class DummyB
+        {
+        }
+
+        public class DummyBDummyFactory : DummyFactory<DummyB>
+        {
+            public static ManualResetEventSlim CreatedLock { get; } = new ManualResetEventSlim(false);
+
+            protected override DummyB Create()
+            {
+                DummyA.CreatingLock.Wait();
+                try
+                {
+                    A.Dummy<DummyA>();
+                }
+                finally
+                {
+                    CreatedLock.Set();
+                }
+
+                return new DummyB();
+            }
+        }
+

```

Comment:
I could not think of a simpler way. But I think I may have done now! I will amend sometime later today.